hail/python/hailtop/utils/gcs_requester_pays.py

```diff
@@ -0,0 +1,25 @@
+from typing import Dict, Optional, Type, TypeVar
+
+from hailtop.aiocloud.aiogoogle import GCSRequesterPaysConfiguration
+
+
+FS = TypeVar("FS")
+
+
+def get_fs_by_requester_pays_config(
+    cls: Type[FS],
+    requester_pays_fses: Dict[Optional[str], FS],
+    requester_pays_config: Optional[GCSRequesterPaysConfiguration] = None
+) -> FS:
+    hashable_config = (
+        requester_pays_config
+        if requester_pays_config is None or isinstance(requester_pays_config, str)
+        else f"{requester_pays_config[0]}:{','.join(requester_pays_config[1])}"
+    )
+    try:
+        return requester_pays_fses[hashable_config]
+    except KeyError:
+        requester_pays_fses[hashable_config] = cls(
+            gcs_kwargs={"gcs_requester_pays_configuration": requester_pays_config}
+        )
```

Comment:
Maybe its the introduction of the hashable stuff, but the bookkeeping around the `requester_pays_fses` is starting to feel a little messy. I wonder if instead of keeping a `dict` around + using this utility method, we could instead [subclass defaultddict](https://stackoverflow.com/questions/43268306/can-defaultdict-accept-callables-dependent-on-the-given-missing-key)?

Reply:
i definitely like the approach of subclassing `DefaultDict` better haha, updated to do that